### ISSUE TITLE
Old compojure dependecy

### ISSUE THREAD

**Issue by Grievoushead** [nginx-clojure/nginx-clojure#182]
`NginxEmbedServer` spins off with old version of compojure, which still has the middleware ordering bug (https://github.com/weavejester/compojure/pull/157).

>nginx-clojure for some reason depends on compojure 1.1.6, which is really ancient.

**Comment by xfeep** [comment]
@Grievoushead Thanks.
nginx-clojure doesn't depend on compojure. Compojure is only used at dev profile for unit tests.

**Comment by Grievoushead** [comment]
@xfeep So then it's only nginx-clojure-embed problem?
I have a problem with `nginx-clojure/nginx-clojure-embed "0.4.5"`: 
Please check this:

 - Create new proj with `lein new compojure-api c2`
 - eplace `project.clj` with

```
(defproject c2 "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [compojure "1.6.0"]
                 [metosin/compojure-api "1.1.10"]]
  :ring {:handler c2.handler/app}
  :uberjar-name "server.jar"
  :profiles {:dev {:dependencies [[javax.servlet/javax.servlet-api "3.1.0"]
                                  [nginx-clojure/nginx-clojure-embed "0.4.5"]]
                   :main c2.handler
                   :plugins [[lein-ring "0.10.0"]]}})
```

 - replace `handler.clj` with

```
(ns c2.handler
  (:require [compojure.api.sweet :refer :all]
            [ring.util.http-response :refer :all]
            [schema.core :as s]
            [nginx.clojure.embed :as embed]))

(s/defschema Collection
  {:id s/Int
   :name s/Str})

(s/defschema Tag
  {:id s/Int
   :tag s/Str})

(def app
  (api
    (swagger-routes)
    (context "/:storeId" []
      (GET "/collections" {:as request params :params}
        :summary "Get Collections"
        :path-params [storeId :- s/Str]
        :return [Collection]
        (ok [{:id 101 :name "My Collection"}]))

      (GET "/tags" {:as request params :params}
        :summary "Get Tags"
        :path-params [storeId :- s/Str]
        :return [Tag]
        (ok [{:id 101 :tag "My Collection"}])))))


(defn start-server
  "Run an emebed nginx-clojure for debug/test usage."
  [dev? port]
  (embed/run-server app {:port port}))


(defn -main
  [& args]
  (let [dev? (empty? args)
        port (or (first args) 8080)
        port (start-server dev? port)]
    (when-not (System/getProperty "java.awt.headless")
      (try
        (.browse (java.awt.Desktop/getDesktop) (java.net.URI. (str "http://localhost:" port "/")))
        (catch java.awt.HeadlessException _)))))
```

 - Now if you run it with `lein with-profile dev run`

The problem will occur, if you try to go to `http://localhost:8080/1/tags` it will throw error since it will try to validate return schema of the first route in same context.
The bug from old compojure https://github.com/weavejester/compojure/pull/157

**Comment by xfeep** [comment]
@Grievoushead Thanks for your example.
I did run `lein deps-tree` with your example project c2 and the result is below

```bash
c2$ lein with-profile dev deps-tree | grep compojure
[compojure "1.6.0"]
[metosin/compojure-api "1.1.10"]
```
So we can see the compojure version is 1.6.0 which is expected.

**Comment by ikitommi** [comment]
Hi. I created a sample minimalstic sample highlighting the problem: https://github.com/ikitommi/nginx-clojure-compojure.

it runs ok with jetty, fails with nginx-clojure...

**Comment by xfeep** [comment]
@ikitommi Thanks!
This issue is caused by nginx-clojure which uses mutable request map and is nothing to do with  old version of compojure.
If we convert the nginx-clojure request to a general clojure map everything will work well. e.g.

```clojure
(def wapp
  (fn[req]
    (app (into {} req))))

(defn start-nginx-clojure [port]
  (embed/run-server wapp {:port port}))
```

**Comment by ikitommi** [comment]
thanks for clarification. I'm bit surprised, could you walk through my example and highlight where & how does the mutable request behave differently? I think it's kind of fundamental if compojure doesn't work with default config.

**Comment by xfeep** [comment]
Technically, it is a bug of nginx-clojure which uses mutable request map instead of clojure persistent map. So operation such as  `assoc` `update-in` will modify the original request which is the same object with the returned.

**Comment by ikitommi** [comment]
thanks. I think this is major. Should the default be changed to use immutable maps? or document  & highlight that the (fast) default is not safe?

**Comment by xfeep** [comment]
The biggest reason of choosing mutable request is not performance but consistence with Java API.
We can get the original request object anywhere regardless how many middlewares are used and we can attach listeners or some useful objects to this request object.
But mutable request really breaks the ring spec. Sorry to see this conflict and will give a tradeoff improvement to the next release.

**Comment by xfeep** [comment]
Fixed in v0.5.0.
